### PR TITLE
hatch: clean, format and fix build

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

cc @onny

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ha/hatch/package.nix [hunk 1]
```diff
@@ -41,51 +47,70 @@ python3.pkgs.buildPythonApplication rec {
...
+  nativeCheckInputs =
+    [ cargo ]
+    ++ (with python3.pkgs; [
+      binary
+      git
+      pytestCheckHook
+      pytest-mock
+      pytest-xdist
+      setuptools
+    ]);
```
Comment (drupol) [inline pkgs/by-name/ha/hatch/package.nix:59]:
```suggestion
  nativeCheckInputs = with python3.pkgs; [
    binary
    git
    pytestCheckHook
    pytest-mock
    pytest-xdist
    setuptools
  ] ++ [ cargo ]
```

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

File: pkgs/by-name/ha/hatch/package.nix [hunk 2]
```diff
@@ -41,51 +47,70 @@ python3.pkgs.buildPythonApplication rec {
...
+      # This test assumes it is running on macOS with a system shell on the PATH.
+      # It is not possible to run it in a nix build using a /nix/store shell.
+      # See https://github.com/pypa/hatch/pull/709 for the relevant code.
+      "test_populate_default_popen_kwargs_executable"
+    ]
+    ++ lib.optionals stdenv.isAarch64 [ "test_resolve" ];
+
+  passthru = {
+    tests.version = testers.testVersion { package = hatch; };
+    updateScript = nix-update-script { };
```
Comment (drupol) [inline pkgs/by-name/ha/hatch/package.nix:105]: Is this really needed? See this comment: https://github.com/NixOS/nixpkgs/pull/334715#discussion_r1720743293

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

File: pkgs/by-name/ha/hatch/package.nix [hunk 3]
```diff
@@ -41,51 +47,70 @@ python3.pkgs.buildPythonApplication rec {
...
+  };
 
-  meta = with lib; {
+  meta = {
```
Comment (drupol) [inline pkgs/by-name/ha/hatch/package.nix:108]: While we are at it, we can also sort those attributes alphabetically.

**Review by drupol — COMMENTED** [review]
Nice refactoring.

Some might complain and require that you split your commit into multiple commits, but for me this is totally OK as it is now.

I also added very little optional comments.

**Review by drupol — APPROVED** [review]

**Comment by GaetanLepage** [comment]
> Some might complain and require that you split your commit into multiple commits, but for me this is totally OK as it is now.
Sure, I tried to do such at the beginning, but then it all got mixed up. I didn't want to go over it again...


I will now try to fix the darwin build, so let's hold on a bit.

**Comment by drupol** [comment]
> Sure, I tried to do such at the beginning, but then it all got mixed up. I didn't want to go over it again...

`git commit -p` and `lazygit` must become your best friends :D

**Comment by GaetanLepage** [comment]
> `git commit -p` and `lazygit` must become your best friends :D

I do use those, but when formatting is applied, it gets very tricky...

File: pkgs/by-name/ha/hatch/package.nix (hunk 3 above)
Comment (GaetanLepage) [reply pkgs/by-name/ha/hatch/package.nix:127]: I changed the order to be the "conventional" one (from what I have seen elsewhere in nixpkgs).
Is that OK for you ?

File: pkgs/by-name/ha/hatch/package.nix (hunk 2 above)
Comment (GaetanLepage) [reply pkgs/by-name/ha/hatch/package.nix:124]: As this one is not under `python312Packages` I would say yes, but maybe I'm wrong.
@natsukium, am I right on this ?

**Comment by GaetanLepage** [comment]
I have skipped the necessary tests to make it build fine on darwin.

File: pkgs/by-name/ha/hatch/package.nix (hunk 3 above)
Comment (drupol) [reply pkgs/by-name/ha/hatch/package.nix:127]: Totally fine. My remark was opinionated. Also, as long as there's no automated tools that enforce this, I can't really ask maintainers to do it.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 335539` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

**Review by drupol — APPROVED** [review]

**Review by onny — APPROVED** [review]

**Comment by onny** [comment]
Result of `nixpkgs-review pr 335539` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

File: pkgs/by-name/ha/hatch/package.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/by-name/ha/hatch/package.nix:124]: It uses a python builder, so it gets the default updater.

Comment (GaetanLepage) [reply pkgs/by-name/ha/hatch/package.nix:124]: Ok, good to know. Thanks !

**Pull request merged**